src/av2/evaluation/detection/utils.py

```diff
@@ -0,0 +1,439 @@
+
+"""Detection utilities for the Argoverse detection leaderboard.
+
+Accepts detections (in Argoverse ground truth format) and ground truth labels
+for computing evaluation metrics for 3d object detection. We have five different,
+metrics: mAP, ATE, ASE, AOE, and CDS. A true positive for mAP is defined as the
+highest confidence prediction within a specified Euclidean distance threshold
+from a bird's-eye view. We prefer these metrics instead of IoU due to the
+increased interpretability of the error modes in a set of detections.
+"""
+
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import numpy as np
+from joblib import Parallel, delayed
+from scipy.spatial.distance import cdist
+
+from av2.evaluation.detection.constants import (
+    MAX_NORMALIZED_ASE,
+    MAX_SCALE_ERROR,
+    MAX_YAW_RAD_ERROR,
+    MIN_AP,
+    MIN_CDS,
+    AffinityType,
+    CompetitionCategories,
+    DistanceType,
+    FilterMetricType,
+    InterpType,
+)
+from av2.geometry.geometry import mat_to_xyz, quat_to_mat, wrap_angles
+from av2.geometry.iou import iou_3d_axis_aligned
+from av2.geometry.se3 import SE3
+from av2.map.map_api import ArgoverseStaticMap, RasterLayerType
+from av2.structures.cuboid import Cuboid, CuboidList
+from av2.utils.constants import EPS
+from av2.utils.io import TimestampedCitySE3EgoPoses, read_city_SE3_ego
+from av2.utils.typing import NDArrayBool, NDArrayFloat, NDArrayInt
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DetectionCfg:
+    """Instantiates a DetectionCfg object for configuring a evaluation.
+
+    Args:
+        affinity_thresholds_m: Affinity thresholds for determining a true positive (in meters).
+        affinity_type: Type of affinity function to be used for calculating average precision.
+        categories: Detection classes for evaluation.
+        eval_only_roi_instances: Only use dets and ground truth that lie within region of interest during eval.
+        filter_metric: Detection metric to use for filtering of both detections and ground truth annotations.
+        max_range_m: Max distance (under a specific metric in meters) for a detection or ground truth cuboid to be
+            considered for evaluation.
+        num_recall_samples: Number of recall points to sample uniformly in [0, 1].
+        splits: Tuple of split names to evaluate.
+        tp_threshold_m: Center distance threshold for the true positive metrics (in meters).
+    """
+
+    affinity_thresholds_m: Tuple[float, ...] = (0.5, 1.0, 2.0, 4.0)
+    affinity_type: AffinityType = AffinityType.CENTER
+    categories: Tuple[str, ...] = tuple(x.value for x in CompetitionCategories)
+    dataset_dir: Optional[Path] = None
+    eval_only_roi_instances: bool = True
+    filter_metric: FilterMetricType = FilterMetricType.EUCLIDEAN
+    max_num_dts_per_category: int = 100
+    max_range_m: float = 200.0
+    num_recall_samples: int = 101
```

Comment:
I think it's cleaner if we use 100 instead and then use `np.linspace(0, 1, cfg.num_recall_samples, **endpoint=True**)`.

Reply:
I really like that --- good call.